File: web/src/lib/qr.ts
// A QR code for the link to a place.
//
// The payload is the permalink, not the code and not a `geo:` URI. Scanning it
// opens this playground at that place, which is the thing being shared. The
// coordinate forms are already in "Also written as" for a reader who wants to
// hand them to a map application instead.

import { renderSVG } from 'uqr';

/**
 * Dark on white, in both themes, and that is deliberate.
 *
 * Everything else on this site follows the reader's theme. This does not: the
 * standard specifies dark modules on a light background, and while many
 * scanners cope with an inverted code, this is a thing whose entire purpose is
 * to be read by a phone nobody here has ever seen. It is a machine-readable
 * target rather than decoration, so it looks the same either way and carries
 * its own white plate.
 *
 * Four modules of quiet zone, which is the minimum the standard asks for. It
 * looks like generous padding and is not: without it, a scanner has no way to
 * find the edge of the symbol.
 *
 * Error correction M, fifteen per cent, because a code on a screen gets
 * photographed at an angle, under a reflection, by a camera that is not quite
 * in focus.
 */
export function qrFor(url: string): string {
    return renderSVG(url, {
        ecc: 'M',
        border: 4,
        pixelSize: 1,
        whiteColor: '#FFFFFF',
        blackColor: '#000000',
    });
}

File: android/core/src/main/kotlin/com/gridpointcode/core/Qr.kt
package com.gridpointcode.core

import com.google.zxing.BarcodeFormat
import com.google.zxing.EncodeHintType
import com.google.zxing.qrcode.QRCodeWriter
import com.google.zxing.qrcode.decoder.ErrorCorrectionLevel

/*
 * A QR code for the link to a place, as the website draws one.
 *
 * The payload is the permalink, directions and all, not the code and not a geo
 * URI: scanning it opens the place, which is the thing being handed over, on any
 * phone, with or without this app. Error correction M, fifteen per cent, because
 * a code on a screen is photographed at an angle, under a reflection, by a
 * camera not quite in focus; and four modules of quiet zone, the least the
 * standard allows, without which a scanner cannot find the symbol's edge.
 */

/** A QR symbol: [size] modules square, quiet zone included, each dark or light. */
class QrSymbol internal constructor(val size: Int, private val dark: BooleanArray) {
    fun isDark(x: Int, y: Int): Boolean = dark[y * size + x]
}

/** Modules of light margin around the symbol, the standard's minimum. */
const val QR_QUIET_ZONE = 4

/** The QR symbol for [link]. */
fun qrFor(link: String): QrSymbol {
    val hints = mapOf(
        EncodeHintType.ERROR_CORRECTION to ErrorCorrectionLevel.M,
        EncodeHintType.MARGIN to QR_QUIET_ZONE,
        EncodeHintType.CHARACTER_SET to "UTF-8",
    )
    // Asked for no size at all, the writer draws one pixel to a module.
    val matrix = QRCodeWriter().encode(link, BarcodeFormat.QR_CODE, 0, 0, hints)
    val size = matrix.width
    return QrSymbol(size, BooleanArray(size * size) { matrix.get(it % size, it / size) })
}
